sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java

```diff
@@ -0,0 +1,393 @@
+                    .flatMap(containerResponse -> rxDocumentClient
+                            .getPartitionKeyRangeCache()
+                            .tryGetOverlappingRangesAsync(
+                                    null,
+                                    containerResponse.getProperties().getResourceId(),
+                                    PartitionKeyInternalHelper.FullRange,
+                                    false,
+                                    null));
+
+            // 1. Extract all preferred read regions to proactively connect to.
+            // 2. Obtain partition addresses for a container for one read region, then mark that read region as unavailable.
+            // 3. This will force resolveAsync to use the next preferred read region in the next invocation.
+            // 4. This way we can verify that connections have been opened to all replicas across all proactive connection regions.
+            for (URI proactiveConnectionEndpoint : proactiveConnectionEndpoints) {
+                Flux.zip(asyncContainerFlux, partitionKeyRangeFlux)
+                        .flatMapIterable(containerToPartitionKeyRanges -> {
+                            List<ImmutablePair<PartitionKeyRange, CosmosAsyncContainer>> pkrToContainer = new ArrayList<>();
+                            for (PartitionKeyRange pkr : containerToPartitionKeyRanges.getT2().v) {
+                                pkrToContainer.add(new ImmutablePair<>(pkr, containerToPartitionKeyRanges.getT1()));
+                            }
+                            return pkrToContainer;
+                        })
+                        .flatMap(partitionKeyRangeToContainer -> {
+                            RxDocumentServiceRequest dummyRequest = RxDocumentServiceRequest.createFromName(
+                                    mockDiagnosticsClientContext(),
+                                    OperationType.Read,
+                                    partitionKeyRangeToContainer.getRight().getLink() + "/docId",
+                                    ResourceType.Document);
+                            dummyRequest.setPartitionKeyRangeIdentity(new PartitionKeyRangeIdentity(partitionKeyRangeToContainer.getLeft().getId()));
+                            return globalAddressResolver.resolveAsync(dummyRequest, false);
+                        })
+                        .delayElements(Duration.ofSeconds(3))
+                        .doOnNext(addressInformations -> {
+                            for (AddressInformation address : addressInformations) {
+                                endpoints.add(address.getPhysicalUri().getURI().getAuthority());
+                            }
+                        })
+                        .blockLast();
+
+                globalEndpointManager.markEndpointUnavailableForRead(proactiveConnectionEndpoint);
+            }
+
+            assertThat(provider.count()).isEqualTo(endpoints.size());
```

Comment:
should these validation also happen before we send request?

Reply:
don't think it adds much value to this test. Won't fix.